src/cc2olx/qti.py

```diff
@@ -0,0 +1,318 @@
+    """
+
+    WEB_RESOURCES_DIR_ALIAS = '$IMS-CC-FILEBASE$'
+    WIKI_CONTENT_DIR_ALIAS = '$WIKI_REFERENCE$'
+
+    def __init__(self, root_xml_doc):
+        self.doc = root_xml_doc
+
+    def create_qti_node(self, details):
+        """
+        Creates OLX xml node, that represents content of unit with problems.
+
+        Args:
+            details: list of dictionaries, where each contains data to
+                render problem.
+        """
+
+        problems = []
+
+        for problem_data in details:
+            cc_profile = problem_data['cc_profile']
+            create_problem = self._problem_creators_map.get(cc_profile)
+
+            if create_problem is None:
+                raise QtiError("Unknown cc_profile: \"{}\"".format(problem_data['cc_profile']))
+
+            problem = create_problem(problem_data)
+            problems.append(problem)
+
+        return problems
+
+    @property
+    def _problem_creators_map(self):
+        """
+        Returns: mapping between Common Cartridge profile value and function
+            that creates actual problem node.
+        """
+        return {
+            MULTIPLE_CHOICE: self._create_multiple_choice_problem,
+            MULTIPLE_RESPONSE: self._create_multiple_response_problem,
+            FILL_IN_THE_BLANK: self._create_fib_problem,
+            ESSAY: self._create_essay_problem,
+            BOOLEAN: self._create_boolean_problem,
+            PATTERN_MATCH: self._create_pattern_match_problem,
+        }
+
+    def _create_multiple_choice_problem(self, problem_data):
+        """
+        Creates XML node of problem.
+        """
+
+        problem = self.doc.createElement("problem")
+        problem_content = self.doc.createElement("choiceresponse")
+
+        problem_description = self._create_problem_description(problem_data['problem_description'])
+
+        choice_group = self.doc.createElement("checkboxgroup")
+
+        for choice_data in problem_data['choices'].values():
+            self._add_choice(
+                choice_group, choice_data['correct'], choice_data['text']
+            )
+
+        problem_content.appendChild(problem_description)
+        problem_content.appendChild(choice_group)
+        problem.appendChild(problem_content)
+
+        return problem
+
+    def _create_problem_description(self, description_html_str):
+        """
+        Material texts can come in form of escaped HTML markup, which
+        can't be considered as valid XML. ``xml.dom.minidom`` has no
+        features to convert HTML to XML, so we use lxml parser here.
+
+        Args:
+            description_html_str: escaped HTML string
+
+        Returns: instance of ``xml.dom.minidom.Node``
+        """
+        description_html_str = unescape(description_html_str)
+
+        # change all image sources to use OLX /static/
+        description_html_str = description_html_str.replace(
+            urllib.parse.quote(self.WEB_RESOURCES_DIR_ALIAS),
```

Comment:
Should we unquote the string rather than quoting the alias to replace here to avoid the incorrect file reference when there's url encoded portions of the filenames?